crates/loam-cli/examples/soroban/nft/src/nft.rs

```diff
@@ -10,10 +10,10 @@ use crate::subcontract::{IsInitable, IsNonFungible};
 pub struct MyNonFungibleToken {
     admin: Address,
     name: Bytes,
-    total_count: u32,
-    owners_to_nft_ids: Map<Address, Map<u32, ()>>, // the owner's collection
-    nft_ids_to_owners: Map<u32, Address>,
-    nft_ids_to_metadata: Map<u32, Bytes>,
+    total_count: u128,
+    owners_to_nft_ids: Map<Address, Map<u128, ()>>, // the owner's collection
+    nft_ids_to_owners: Map<u128, Address>,
+    nft_ids_to_metadata: Map<u128, Bytes>,
```

Comment:
Hmm it seems annoying that the metadata is untyped. Would be nice for consumers of the contract to know what the shape of the metadata is. But if is the current standard it's fine, but it's something we should push on to get agreement.

Reply:
I assumed that this was an intentional part of the spec, to allow rapid ecosystem experimentation and _de facto_ standards to arise.

Reply:
Yeah, that's a great point. From what I can tell from [SEP-0039](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0039.md#describing-your-nft) best practices are to include a few key fields in the metadata: name, description, URL, issue, and code. 

What if we start with those fields in a Metadata struct, and then if/when a standard emerges we can update the type? Not sure how difficult it will be to upgrade subcontracts, though with the ability to redeploy, that seems doable.
```
pub struct Metadata {
    name: Bytes,
    description: Bytes,
    url: Bytes,
    issuer: Address,
    code: Bytes,
}
```

alternatively we could have them all be optional fields? 🤔

Reply:
Personally I think the first three should be required. And Perhaps we use `String`?. It's again annoying because the all seem like that is what they will be. So perhaps we just put this forward as our standard. Other contracts will convert from Bytes to string anyway so we are just skipping a step.

Reply:
Could we implement it as two separate subcontracts, `IsSep39` and `IsSep39Extension`, or something like that? Implementing `IsSep39Extension` would override all instances of metadata from `Bytes` to this suggested `Metadata` struct.

Reply:
Or `IsSep39WithTypedMetadata`, or `IsTypedMetadata`, or... ¯\\\_(ツ)_/¯

Reply:
Finally getting back to addressing this. 

I like the idea of creating an `IsSep39Metadata` subcontract, but I'm unsure how to implement it. 

- One way I tried was to have an `IsSep39Metdata` subcontract with a `Metadata` associated type:
```
impl IsTypedMetadata for MyNonFungibleToken {
    type Metadata = MyNonFungibleTokenMetadata;
}
```
The issue I ran into with this is that our `derive_contract` doesn't generate the associate type over to the generated trait `TypedMetadata`. It just adds the `Impl` generated type. I'm not sure if this is something we actually want to add to the macro, or if there is a better way to do this, but I moved on and tried a different idea...

- The other idea was to have `IsTypedMetadata` include a `deserialize_metadata` fn that transforms the `Bytes` metadata into the MyNonFungibleTokenMetadata type. But that is still annoying because the user needs to serialize the metadata into Bytes when minting, etc. 

I'll keep thinking on this, but I could probably use another set of eyes on this.

Reply:
Also, for now I've just added a Metadata type like the following, maybe we can move forward with that and then continue to improve the standard/make sure that the Metadata can be flexible going forward.

```
pub struct Metadata {
    pub(crate) name: String,
    pub(crate) description: String,
    pub(crate) url: String,
}
```

Reply:
With `#[contracttype]` the name of the type exported, so perhaps we could just called it `Sep39Metadata`?